compiler/balt.bal

```diff
@@ -0,0 +1,162 @@
+                // possible improvement: support other white spaces
+                if l.startsWith(" ") {
+                    if prevFiledName == () ||  prevFiledBody == () {
+                        panic err:impossible("folded field body without field");
+                    }
+                    header[<string>prevFiledName] = <string>prevFiledBody + l;
+                } else {
+                    var [fName, fBody] = parseField(l);
+                    header[fName] = fBody;
+                    prevFiledName = fName;
+                    prevFiledBody = fBody;
+                }
+            }
+        }
+        else if s == CONTENT {
+            content.push(l);
+        }
+        else if s == BOF {
+            // xxx add file path
+            return err:syntax("file should start with 'Test-Case:' header field");
+        }
+        else {
+            panic err:impossible("balt parser illegal state");
+        }
+    }
+    if s == HEADER {
+        return err:syntax("header without content at EOF ");
+    }
+    else if s == BOF {
+        return err:syntax("empty file");
+    }
+    else {
+        BaltTestHeader header = <BaltTestHeader>maybeHeader;
+        tests.push({offset, header, content});
+    }
+    return tests;
+}
+
+function parseField(string s) returns [string, string] {
+    int? i = s.indexOf(":");
+    if i == () {
+        panic error("Missing `:` in field : " + s);
+    } else {
+        string fieldName = s.substring(0, i);
+        string fieldValue = s.substring(i + 1);
+        return [fieldName, fieldValue.trim()];
+    }
+}
+
+const OUTPUT_MARKER = "@output";
+const ERROR_MARKER = "@error";
+const PANIC_MARKER = "@panic";
+
+
+function expect(string[] src) returns string[] {
+    string[] expect = [];
+    foreach var [i, l] in src.enumerate() {
+        int? comment = l.indexOf("//");
+        if comment == () {
+            continue;
+        }
+
+        int begin = <int> comment + 2;
+        int? output = l.indexOf(OUTPUT_MARKER, begin);
+        if output != () {
+            expect.push(l.substring(output + OUTPUT_MARKER.length()).trim());
+        }
+        int? err = l.indexOf(ERROR_MARKER, begin);
+        if err != () {
+            expect.push(ERROR_MARKER);
+        }
+        int? pnk = l.indexOf(PANIC_MARKER, begin);
+        if pnk != () {
+            string msg = l.substring(pnk + PANIC_MARKER.length()).trim();
+            expect.push("panic: line " + (i + 1).toString() + ": " + msg);
+        }
+    }
+    return expect;
+}
+
+const FNV1_32_INIT = 0x811c9dc5;
+const FNV1_32_PRIME = 0x1000193;
+
+function hash(string[] lines) returns int {
+    int hash = FNV1_32_INIT;
+    foreach var l in lines {
+        hash = fnv1a32(l.toBytes(), hash);
+    }
+    return hash;
+}
+
+function fnv1a32(byte[] bs, int init = FNV1_32_INIT) returns int {
+    int hash = init;
+    foreach byte b in bs {
+        hash ^= b;
+        hash *= FNV1_32_PRIME;
```

Comment:
Isn't this going to overflow?

Reply:
No, because hash is at most `0xffffffff`